query-algebrizer/src/cc.rs

```diff
@@ -209,14 +213,31 @@ pub struct ConjoiningClauses {
     pub wheres: Vec<ColumnConstraint>,
 
     /// A map from var to qualified columns. Used to project.
-    pub bindings: BTreeMap<Variable, Vec<QualifiedAlias>>,
+    pub column_bindings: BTreeMap<Variable, Vec<QualifiedAlias>>,
+
+    /// A list of variables mentioned in the query's :in clause. These must all be bound before the
+    /// query can be executed.
+    pub input_variables: BTreeSet<Variable>,
+
+    /// In some situations -- e.g., when a query is being run only once -- we know the values of
+    /// some or all bindings in advance. These can be substituted directly when the query is
+    /// algebrized.
+    ///
+    /// Known bindings must agree with `known_types`. If you write a query like
+    /// ```edn
+    /// [:find ?x :in $ ?val :where [?x :foo/int ?val]]
+    /// ```
+    ///
+    /// and for `?val` provide `TypedValue::String("foo".to_string())`, the query will be known at
+    /// algebrizing time to be empty.
+    known_bindings: BTreeMap<Variable, TypedValue>,
```

Comment:
You might consider calling this `input_values`.  Is there ever a time that `input_variables` will not be a superset of the keys of `known_bindings`?  (I may answer this question as I read.)

You might also consider `inputs: BTreeMap<Variable, Option<TypedValue>>` to unify where you track things.  Perhaps this doesn't make sense for data sources (i.e., for `$`).